Split options and flags in kubectl config sub-commands

mpuckett159 opened the pull request:
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
Part of ongoing effort to split options and flags out in kubectl commands.

#### Special notes for your reviewer:

The way I have put this together we have three main categories of functions:

Commands that takes flags
- New\<subcommand\>Flags - This function creates a new initialized set of flags passing in the config file access and io streams objects at a minimum. All other flags are granted default values here.
- AddFlags - This function adds the flags to the `cobra.Command` object.
- ToOptions - This function takes the arguments provided by the user, the flags provided by the user, the default values of the flags, and composes all of this into a set of subcommand options that are returned which will then be later used by the run command.

Commands that do not take flags
- New\<subcommand\>Options - This function creates a new initialized set of options, passing the config file access and io streams objects at a minimum.
- Complete - This function takes the arguments provided by the user that would normally be added to the options object in the ToOptions function and adds them to the options object.

Utility functions
- Validate - Sometimes it is necessary to apply complicated logic to verify that an options object is both complete and valid based on user inputs. This function can accomplish this step if necessary.

#### Does this PR introduce a user-facing change?
No

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
See following issue for more details.

https://github.com/kubernetes/kubectl/issues/1046

mpuckett159 commented:
Alright as of this comment the subcommands in the config subcommand that use flags have been reworked to conform to the split flags/options model laid out in the wait subcommand. The list of changed files in the PR will tell you which that is.

All config subcommand tests pass, however, the config_test.go tests do not due to the way the tests are being run, not due to any underlying instability in the code itself. I'm not 100% sure but I think I need to rewrite the tests completely for it because of the method being used to ignore returned errors, which effectively allows the command to continue on error. This results in a null pointer error due to the new flags to options process.

Anyway, just fyi if people are reviewing before I get that fixed up.

seans3 commented:
/ok-to-test

mpuckett159 commented:
Ok turns out not all tests pass sans `config_test.go` tests, easy fix though.

mpuckett159 commented:
I believe integration is just a flake, but not going to retest until I get the `config_test.go` junk cleaned up.

mpuckett159 commented:
/retest

mpuckett159 commented:
Well that's strange all the unit tests pass for me locally. Back to the drawing board.

mpuckett159 commented:
/retest

mpuckett159 commented:
Oh good my favorite, do nothing and the problem goes away.

So now that I have a fix for the commands that required rewritting some of their logic, I'm gonna clean up the rest, then clean up the commit history and tag this as ready for review.

mpuckett159 commented:
Still not fully done but mostly done, really need to get the set commands redone and standardized, and then dig into getting the config tests done a bit better.

mpuckett159 commented:
/remove-lifecycle stale

dims commented:
If you still need this PR then please rebase, if not, please close the PR

dims commented:
This PR has the label work-in-progress, please revisit to see if you still need this, please close if not

mpuckett159 commented:
I know these are probably automated messages from dims but for clarity this WIP is still (intending to be) worked on and completed. I have my notes on what needs to be updated and fixed and just need to finish the work.

mpuckett159 committed:
Add common testing funcs

mpuckett159 committed:
Add loadConfig func

Reused very frequently and very common function call. This should make it very easy to load a config without having to copy paste around 10 lines of code everywhere

mpuckett159 committed:
Remove helpErrorf func

Only used in current-context and does not fit the general pattern of the other kc config commands

mpuckett159 committed:
Rework current-context

mpuckett159 committed:
Rework delete-cluster

mpuckett159 committed:
Rework delete-context

mpuckett159 committed:
Rework delete-user

mpuckett159 committed:
Rework get-clusters

mpuckett159 committed:
Rework get-contexts

mpuckett159 committed:
Rework get-users

mpuckett159 committed:
Rework rename-context

mpuckett159 committed:
Rework set

mpuckett159 committed:
Rework set-cluster

mpuckett159 committed:
Rework set-context

mpuckett159 committed:
Rework set-credentials

mpuckett159 committed:
Rework unset

mpuckett159 committed:
Rework use-context

mpuckett159 committed:
Rework view

mpuckett159 committed:
Rework config

mpuckett159 commented:
/ok-to-test

mpuckett159 committed:
fix verify errors

mpuckett159 commented:
/retest

mpuckett159 commented:
/test pull-kubernetes-integration

mpuckett159 commented:
/test pull-kubernetes-unit

mpuckett159 commented:
/release-note-none
/priority backlog
/triage accepted

I think I need to stop tweeking this an just explain the reasoning.

mpuckett159 commented:
/test pull-kubernetes-conformance-kind-ga-only-parallel

mpuckett159 commented:
This should almost certainly have been split into smaller PRs sorry to whoever ends up reviewing this 😬

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/config/config_test.go:
```diff
@@ -364,24 +371,24 @@ func TestTokenAndBasicDisallowed(t *testing.T) {
-		})
-
+	// Allow for command errors in testing
+	if os.Getenv("CRASH") == "1" {
```
I did not know that environment variable, is it something widely used in k8s and what does it for?

mpuckett159 replied:
This is a way to allow for commands to issue OS exit codes when they fail and keep testing them. Go doesn't really advise doing this, but it's possible, and I was struggling with the method used before for unit tests in these files, so I converted them to this method. What I intend to do is to just move these kinds of tests to the integration testing because that is generally what is advised these days for testing failure states like this. I'll need to do that in a separate PR as noted in my other comment.

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/config/current_context.go:
```diff
@@ -42,35 +37,56 @@ var (
 // RunCurrentContext performs the execution of 'config current-context' sub command
-func RunCurrentContext(out io.Writer, options *CurrentContextOptions) error {
-	config, err := options.ConfigAccess.GetStartingConfig()
+func (o *CurrentContextOptions) RunCurrentContext() error {
```
Maybe now we can safely rename this function to `Run`, since we already know that this is `CurrentContextOptions` function?

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/config/current_context.go:
```diff
@@ -42,35 +37,56 @@ var (
 	}
 
-	fmt.Fprintf(out, "%s\n", config.CurrentContext)
+	if _, err := fmt.Fprintf(o.IOStreams.Out, "%s\n", config.CurrentContext); err != nil {
```
How this function can get error?, even if it gets somehow an error, should it block the command execution?

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/config/delete_cluster.go:
```diff
@@ -44,43 +53,50 @@ func NewCmdConfigDeleteCluster(out io.Writer, configAccess clientcmd.ConfigAcces
-	configFile := configAccess.GetDefaultFilename()
-	if configAccess.IsExplicitFile() {
-		configFile = configAccess.GetExplicitFile()
+func (o *DeleteClusterOptions) RunDeleteCluster() error {
```
Same above and all the others.

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/config/delete_context.go:
```diff
@@ -44,47 +53,56 @@ func NewCmdConfigDeleteContext(out, errOut io.Writer, configAccess clientcmd.Con
 	if len(args) != 1 {
-		cmd.Help()
-		return nil
+		return fmt.Errorf("unexpected args: %v", args)
```
Isn't this changing the behavior?

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/config/delete_user.go:
```diff
@@ -38,27 +37,14 @@ var (
 // DeleteUserOptions holds the data needed to run the command
 type DeleteUserOptions struct {
-	user string
+	User string
```
What is the reason of exporting this field?

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/config/delete_user.go:
```diff
@@ -67,56 +53,51 @@ func NewCmdConfigDeleteUser(streams genericclioptions.IOStreams, configAccess cl
-		return fmt.Errorf("cannot delete user %s, not in %s", o.user, o.configFile)
+// RunDeleteUser performs the command
+func (o *DeleteUserOptions) RunDeleteUser() error {
+	config, configFile, err := loadConfig(o.ConfigAccess)
```
Is this refactoring related to the options and flags split?

ardaguclu reviewed (commented):
Thanks for the PR, this is a good effort and generally all looks good.

I think, there are different kind of changes are all baked in this PR. Would it be better to open separate PRs for each of them and this PR only includes options->flag split?

Because, we need to ensure that unit tests are verifying this changes but in this PR unit tests are also being changed inevitably. Also would you mind adding more integrations tests to assure that everything works as expected.

mpuckett159 commented:
Yeah, you know, honestly, you're right. I let this one get away from me. I'm going to break this up into smaller PRs. This thing is a nightmare. I let the scope creep way too far.
